state-chain/pallets/cf-elections/src/electoral_systems/liveness.rs

```diff
@@ -0,0 +1,183 @@
+		electoral_access: &mut ElectoralAccess,
+		election_identifiers: Vec<ElectionIdentifierOf<Self>>,
+		(current_sc_block, current_chain_tracking_number): &Self::OnFinalizeContext,
+	) -> Result<Self::OnFinalizeReturn, CorruptStorageError> {
+		fn block_number_to_check<ChainBlockNumber: Into<u64> + From<u64>>(
+			current_chain_tracking_number: ChainBlockNumber,
+		) -> ChainBlockNumber {
+			use nanorand::{Rng, WyRand};
+			let height: u64 = current_chain_tracking_number.into();
+
+			// 100 seems a reasonable number for all chains, fast or slow, most running nodes will
+			// have at least 100 blocks worth of state for any particular chain, while still
+			// providing enough variation.
+			const RANGE: u64 = 100;
+			WyRand::new_seed(height)
+				.generate_range((height.saturating_sub(RANGE))..height)
+				.into()
+		}
+
+		if let Some(election_identifier) = election_identifiers
+			.into_iter()
+			.at_most_one()
+			.map_err(|_| CorruptStorageError::new())?
+		{
+			let mut election_access = electoral_access.election_mut(election_identifier)?;
+
+			// Is the block the election started at + the duration we want the check to stay open
+			// for less than or equal to the current SC block?
+			if election_access.state()?.saturating_add(election_access.settings()?) <=
+				*current_sc_block
+			{
+				if let Some(bad_validators) = election_access.check_consensus()?.has_consensus() {
+					if !bad_validators.is_empty() {
+						Hook::on_check_complete(bad_validators);
+					}
+				}
+				election_access.delete();
+				electoral_access.new_election(
+					(),
+					block_number_to_check(*current_chain_tracking_number),
+					*current_sc_block,
+				)?;
+			}
+		} else {
+			electoral_access.new_election(
+				(),
+				block_number_to_check(*current_chain_tracking_number),
+				*current_sc_block,
+			)?;
+		}
+
+		Ok(())
+	}
+
+	fn check_consensus<ElectionAccess: ElectionReadAccess<ElectoralSystem = Self>>(
+		_election_identifier: ElectionIdentifierOf<Self>,
+		_election_access: &ElectionAccess,
+		_previous_consensus: Option<&Self::Consensus>,
+		consensus_votes: ConsensusVotes<Self>,
+	) -> Result<Option<Self::Consensus>, CorruptStorageError> {
+		let num_authorities = consensus_votes.num_authorities();
+		let success_threshold = success_threshold_from_share_count(num_authorities);
+
+		let mut active_votes = BTreeMap::new();
+		let mut non_voters = Vec::new();
+		for ConsensusVote { vote, validator_id } in consensus_votes.votes {
+			match vote {
+				Some(vote) => {
+					active_votes.entry(vote.1).or_insert_with(Vec::new).push(validator_id);
+				},
+				None => non_voters.push(validator_id),
+			}
+		}
+
+		let (consensus_validators, non_consensus_validators): (Vec<_>, Vec<_>) =
+			active_votes.into_iter().partition(|(_, validator_ids)| {
+				validator_ids.len() as AuthorityCount >= success_threshold
+			});
+
+		debug_assert!(
+			consensus_validators.len() <= 1,
+			"there should only be one group that gained consensus"
+		);
```

Comment:
We know this is true because success_threshold is a majority if votes. So since BTreeMap is sorted, we could just take the last entry and check if it contains more than success_threshold votes.

Reply:
Removed as part of making it a single map